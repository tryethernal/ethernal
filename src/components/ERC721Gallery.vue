<template>
    <v-container fluid>
        <template v-if="has721Enumerable && parseInt(totalSupply) > 0">
            <v-row>
                <v-col cols="6" sm="4" lg="2" v-for="(token, idx) in tokens" :key="idx">
                    <ERC721-Token-Card
                        :owner="token.owner"
                        :name="token.attributes.name"
                        :imageData="token.attributes.image_data"
                        :index="token.index"
                        :tokenId="token.tokenId"
                        :contractAddress="address"
                        :backgroundColor="token.attributes.background_color" />
                </v-col>
            </v-row>
            <v-row>
                <v-col>
                    <v-pagination
                        v-model="page"
                        :length="length"
                        :total-visible="7"
                        @input="pageChanged">
                    </v-pagination>
                </v-col>
            </v-row>
        </template>
        <template v-else-if="!has721Enumerable">
            <v-card outlined>
                <v-card-text>
                    Collection view is only available for ERC721 contracts that implement the Enumerable extension.
                </v-card-text>
            </v-card>
        </template>
        <template v-else-if="has721Enumerable && parseInt(totalSupply) == 0">
            <v-card outlined>
                <v-card-text>
                    It looks like this collection is empty. Mint some tokens to see them (if you just minted some, wait a few more seconds for them to be processed).
                </v-card-text>
            </v-card>
        </template>
        <template v-else>
            <v-card outlined>
                <v-card-text>
                    Error while displaying NFT collection.
                </v-card-text>
            </v-card>
        </template>
    </v-container>
</template>

<script>
import { mapGetters } from 'vuex';
import ERC721TokenCard from './ERC721TokenCard';

export default {
    name: 'ERC721Gallery',
    props: ['address', 'totalSupply', 'has721Enumerable'],
    components: {
        ERC721TokenCard
    },
    data: () => ({
        page: 1,
        tokens: [],
        currentOptions: { page: 1, itemsPerPage: 12, sortBy: ['index'], order: 'asc' }
    }),
    mounted() {
        this.getTokens();
    },
    methods: {
        pageChanged(newPage) {
            this.currentOptions = { ...this.currentOptions, page: newPage };
            this.getTokens();
        },
        getTokens() {
            this.tokens = Array(this.maxTokenLength).fill({ attributes: {}});
            this.server.getErc721Tokens(this.address, this.currentOptions, this.currentWorkspace.erc721LoadingEnabled)
                .then(({ data: { items } }) => {
                    const tokens = [];
                    for (let i = 0; i < this.maxTokenLength; i++)
                        tokens.push(items[i] ? items[i] : { attributes: {}});
                    this.tokens = tokens;
                })
                .catch(console.log)
        }
    },
    computed: {
        ...mapGetters([
            'currentWorkspace'
        ]),
        maxTokenLength() {
            if (this.currentOptions.page == this.length)
                return this.currentOptions.itemsPerPage - (this.length * this.currentOptions.itemsPerPage - this.totalSupply);
            else
                return this.currentOptions.itemsPerPage
        },
        length() {
            return Math.ceil(this.totalSupply / this.currentOptions.itemsPerPage);
        }
    }
}
</script>
