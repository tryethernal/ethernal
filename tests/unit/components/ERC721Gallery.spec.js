import MockHelper from '../MockHelper';
import flushPromises from 'flush-promises'
import ERC721Gallery from '@/components/ERC721Gallery.vue';

const helper = new MockHelper();
const stubs = [
    'ERC721-Token-Card'
];

describe('ERC721Gallery.vue', () => {
    beforeEach(() => jest.clearAllMocks());

    it('Should load & display tokens', async () => {
        jest.spyOn(helper.mocks.server, 'getErc721Tokens')
            .mockResolvedValue({ data: { total: 1, items: [{ tokenId: '1', owner: '0xabc', attributes: { name: 'Ethernal' }}]}});
        
        const wrapper = helper.mountFn(ERC721Gallery, {
            propsData: {
                address: 'Ox123',
                totalSupply: 100,
                has721Enumerable: true
            },
            stubs: stubs
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display info message if contract is not enumerable', async () => {
        jest.spyOn(helper.mocks.server, 'getErc721Tokens')
            .mockResolvedValue({ data: { total: 1, items: [{ tokenId: '1', owner: '0xabc', attributes: { name: 'Ethernal' }}]}});
        
        const wrapper = helper.mountFn(ERC721Gallery, {
            propsData: {
                address: 'Ox123',
                totalSupply: 100,
                has721Enumerable: false
            },
            stubs: stubs
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display info message if contract is enumerable but there are no tokens minted', async () => {
        jest.spyOn(helper.mocks.server, 'getErc721Tokens')
            .mockResolvedValue({ data: { total: 0, items: []}});
        
        const wrapper = helper.mountFn(ERC721Gallery, {
            propsData: {
                address: 'Ox123',
                totalSupply: 0,
                has721Enumerable: true
            },
            stubs: stubs
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});