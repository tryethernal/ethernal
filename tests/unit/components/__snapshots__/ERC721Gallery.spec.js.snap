// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ERC721Gallery.vue Should display info message if contract is enumerable but there are no tokens minted 1`] = `
<div class="container container--fluid">
  <div class="v-card v-sheet v-sheet--outlined theme--light">
    <div class="v-card__text">
      It looks like this collection is empty. Mint some tokens to see them (if you just minted some, wait a few more seconds for them to be processed).
    </div>
  </div>
</div>
`;

exports[`ERC721Gallery.vue Should display info message if contract is not enumerable 1`] = `
<div class="container container--fluid">
  <div class="v-card v-sheet v-sheet--outlined theme--light">
    <div class="v-card__text">
      Collection view is only available for ERC721 contracts that implement the Enumerable extension.
    </div>
  </div>
</div>
`;

exports[`ERC721Gallery.vue Should load & display tokens 1`] = `
<div class="container container--fluid">
  <div class="row">
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123" name="Ethernal" tokenid="1" owner="0xabc"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
    <div class="col-sm-4 col-lg-2 col-6">
      <erc721-token-card-stub contractaddress="Ox123"></erc721-token-card-stub>
    </div>
  </div>
  <div class="row">
    <div class="col">
      <nav role="navigation" aria-label="Pagination Navigation">
        <ul class="v-pagination theme--light">
          <li><button disabled="disabled" type="button" aria-label="Previous page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
          <li><button type="button" aria-current="true" aria-label="Current Page, Page 1" class="v-pagination__item v-pagination__item--active primary">1</button></li>
          <li><button type="button" aria-label="Goto Page 2" class="v-pagination__item">2</button></li>
          <li><button type="button" aria-label="Goto Page 3" class="v-pagination__item">3</button></li>
          <li><span class="v-pagination__more">...</span></li>
          <li><button type="button" aria-label="Goto Page 7" class="v-pagination__item">7</button></li>
          <li><button type="button" aria-label="Goto Page 8" class="v-pagination__item">8</button></li>
          <li><button type="button" aria-label="Goto Page 9" class="v-pagination__item">9</button></li>
          <li><button type="button" aria-label="Next page" class="v-pagination__navigation"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
